spamc.go

```diff
@@ -89,23 +228,56 @@ func (c *Client) write(
 	w := bufio.NewWriter(buf)
 	tp := textproto.NewWriter(w)
 
-	err := tp.PrintfLine("%v SPAMC/%v", cmd, clientProtocolVersion)
-	if err != nil {
+	// Make sure to seek to start of message
+	if _, err := message.Seek(0, io.SeekStart); err != nil {
 		return err
 	}
 
-	// Make the sure message always ends in \r\n, if it doesn't SA will just
-	// keep reading from the connection and it will block until it timouts.
-	if !strings.HasSuffix(message, "\r\n") {
-		message += "\r\n"
+	size := int64(-1)
+	switch v := message.(type) {
+	case *strings.Reader:
+		size = v.Size()
+	case *bytes.Reader:
+		size = v.Size()
+	case *os.File:
+		stat, err := v.Stat()
+		if err != nil {
+			return err
+		}
+		size = stat.Size()
 	}
 
-	// Always add Content-length header.
-	err = tp.PrintfLine("Content-length: %v", len(message))
+	// make sure to delete a present content-length (and use it if
+	// we could not set size above
+	if _, ok := headers[HeaderContentLength]; ok {
+		if size == -1 {
+			headerContentLength, err := strconv.ParseInt(headers.Get(HeaderContentLength), 10, 64)
+			if err != nil {
+				return err
+			}
+			size = headerContentLength
+		}
+	}
+
+	// Make the sure message always ends in \r\n, if it doesn't SA will just
+	// keep reading from the connection and it will block until it timeouts.
+
+	// messageReader contains the new message messageSize and we suppress warnings, _
+	messageReader, messageSize, _ := c.rearrangeMessageReader(message, size)
```

Comment:
I don't think we want to ignore the error here?

Reply:
Well it's a choice really .. if sending say the message 'Message' only it will generate a "warning" .. it will not "crash or break per se" as Message is the 1st line and does not contain a valid key:value pair it will generate a 'malformed MIME header line' warning and will insert a \r\n before this 'Message' (and will result in '\r\nMessage\r\n') ..
I could also refactor to remove returning the errors, but we would loose future possibilities to "log" these warnings.

Reply:
If we could have some logging build into the code it would be best to log these warnings.  They are warnings really, if we were to treat them as errors and error out it would prevent us passing any arbitrary string for SA to scan, sending 'Message' turns into '\r\nMessage\r\n' with a 'malformed MIME header line:' warning but that does not (in the current implementation) imply that this is wrong, just that the caller choose not to send any headers and choose to new line terminate the body, erroring out would force the user to pass 'email like messages only'